fda/grid.py

```diff
@@ -543,91 +547,151 @@ def concatenate(self, other):
                          self.dataset_label,
                          self.axes_labels)
 
-    def _set_labels(self, ax):
+    def _set_labels(self, fig):
+        # TODO: check in init the length of the labels
         """Set labels if any.
 
         Args:
-            ax (axes object): axes object that implements set_title,
-                set_xlable and set_ylabel or title, xlabel and ylabel.
+            fig (figure object): figure object containing the axes that implement set_xlabel and set_ylabel,
+                                 and set_zlabel in case of a 3d projection.
 
         """
         if self.dataset_label is not None:
-            try:
-                ax.set_title(self.dataset_label)
-            except AttributeError:
-                try:
-                    ax.title(self.dataset_label)
-                except AttributeError:
-                    pass
+            fig.suptitle(self.dataset_label, fontsize="x-large")
 
         if self.axes_labels is not None:
-            try:
-                ax.set_xlabel(self.axes_labels[0])
-                ax.set_ylabel(self.axes_labels[1])
-            except AttributeError:
-                try:
-                    ax.xlabel(self.axes_labels[0])
-                    ax.ylabel(self.axes_labels[1])
-                except AttributeError:
-                    pass
-
-    def plot(self, ax=None, **kwargs):
+            ax = fig.get_axes()
+            if ax[0].name == '3d':
+                for i in range(self.ndim_image):
+                    ax[i].set_xlabel(self.axes_labels[0])
+                    ax[i].set_ylabel(self.axes_labels[1])
+                    ax[i].set_zlabel(self.axes_labels[i + 2])
+            else:
+                for i in range(self.ndim_image):
+                    ax[i].set_xlabel(self.axes_labels[0])
+                    ax[i].set_ylabel(self.axes_labels[i + 1])
+
+    def _set_figure_and_axes(self):
+        """Set the figure and its axes.
+
+        Returns:
+            fig (figure object): figure object initialiazed.
+            ax (axes object): axes of the initialized figure.
+
+        """
+        fig = plt.figure()
+
+        if self.ndim_domain == 1:
+            projection = None
+        else:
+            projection = '3d'
+
+        ncols = math.ceil(math.sqrt(self.ndim_image))
+        nrows = math.ceil(self.ndim_image / ncols)
+        for i in range(self.ndim_image):
+            fig.add_subplot(nrows, ncols, i + 1, projection=projection)
+
+        ax = fig.get_axes()
+
+        return fig, ax
+
+    def _arrange_layout(self, fig):
+        """Arrange the layout of the figure.
+
+        Args:
+            fig (figure object): figure object to be arranged.
+
+        """
+        fig.tight_layout()
+        if self.dataset_label is not None:
+            st = fig.texts[0]
+            st.set_y(0.95)
```

Comment:
I am not sure about the magic numbers. Why are they necessary?

Reply:
They are necessary because the figure is composed of several axes objects without taking into account if we want a common title for all of them (dataset_label). If this was not done, the title would  overlap with the axes on top.